File: Python/Array/M-452. Minimum Number of Arrows to Burst Balloons.py
class Solution:
    @classmethod
    def findMinArrowShots(self, points: list[list[int]]) -> int:
        points.sort(key=lambda a: a[1])
        min_arrow = 1
        prev_idx = 0
        for i in range(1, len(points)):
            _, prev_end = points[prev_idx]
            start, _ = points[i]
            if start <= prev_end:
                continue
            min_arrow += 1
            prev_idx = i
        return min_arrow


testData = [
    {
        "points": [
            [10, 16],
            [2, 8],
            [1, 6],
            [7, 12],
        ],
        "ans": 2,
    },
    {
        "points": [
            [1, 2],
            [3, 4],
            [5, 6],
            [7, 8],
        ],
        "ans": 4,
    },
    {
        "points": [
            [1, 2],
            [2, 3],
            [3, 4],
            [4, 5],
        ],
        "ans": 2,
    },
]

for data in testData:
    print(data["ans"] == Solution.findMinArrowShots(data["points"]))

File: Array/M-452. Minimum Number of Arrows to Burst Balloons/M-452. Minimum Number of Arrows to Burst Balloons.ts
function findMinArrowShots(points: number[][]): number {
  points.sort((a, b) => a[1] - b[1]);
  let prevIdx = 0;
  let minArrow = 1;
  for (let i = 1; i < points.length; i++) {
    const [_prevStart, prevEnd] = points[prevIdx];
    const [start, _end] = points[i];
    if (start <= prevEnd) continue;
    minArrow++;
    prevIdx = i;
  }
  return minArrow;
}

const testData = [
  {
    points: [
      [10, 16],
      [2, 8],
      [1, 6],
      [7, 12],
    ],
    ans: 2, // 6, 11
  },
  {
    points: [
      [1, 2],
      [3, 4],
      [5, 6],
      [7, 8],
    ],
    ans: 4,
  },
  {
    points: [
      [1, 2],
      [2, 3],
      [3, 4],
      [4, 5],
    ],
    ans: 2, //2, 4
  },
];

for (const data of testData) {
  console.log(data['ans'] === findMinArrowShots(data['points']));
}
